dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/export/enrollment/JdbcEnrollmentStore.java

```diff
@@ -0,0 +1,641 @@
+    }
+  }
+
+  private void addOrgUnitConditions(
+      StringBuilder sql,
+      EnrollmentQueryParams params,
+      MapSqlParameterSource sqlParams,
+      SqlHelper hlp) {
+    if (params.hasOrganisationUnits()) {
+      if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.DESCENDANTS)) {
+        sql.append(hlp.whereAnd())
+            .append(getDescendantsQuery(params.getOrganisationUnits(), sqlParams));
+      } else if (params.isOrganisationUnitMode(OrganisationUnitSelectionMode.CHILDREN)) {
+        sql.append(hlp.whereAnd())
+            .append(getChildrenQuery(params.getOrganisationUnits(), sqlParams));
+      } else {
+        sql.append(hlp.whereAnd()).append("en_ou.uid IN (:orgUnitUids)");
+        sqlParams.addValue("orgUnitUids", getUids(params.getOrganisationUnits()));
+      }
+    }
+  }
+
+  private void addProgramConditions(
+      StringBuilder sql,
+      EnrollmentQueryParams params,
+      MapSqlParameterSource sqlParams,
+      SqlHelper hlp) {
+    sql.append(hlp.whereAnd()).append("p.type = :programType");
+    sqlParams.addValue("programType", ProgramType.WITH_REGISTRATION.name());
+
+    if (params.hasProgram()) {
+      sql.append(hlp.whereAnd()).append("p.uid = :programUid");
+      sqlParams.addValue("programUid", params.getProgram().getUid());
+    }
+
+    if (params.hasProgramStartDate()) {
+      sql.append(hlp.whereAnd()).append("e.enrollmentdate >= :programStartDate");
+      sqlParams.addValue("programStartDate", new Timestamp(params.getProgramStartDate().getTime()));
+    }
+
+    if (params.hasProgramEndDate()) {
+      sql.append(hlp.whereAnd()).append("e.enrollmentdate <= :programEndDate");
+      sqlParams.addValue("programEndDate", new Timestamp(params.getProgramEndDate().getTime()));
+    }
+  }
+
+  private void addEnrollmentConditions(
+      StringBuilder sql,
+      EnrollmentQueryParams params,
+      MapSqlParameterSource sqlParams,
+      SqlHelper hlp) {
+    if (params.hasEnrollmentUids()) {
+      sql.append(hlp.whereAnd()).append("e.uid in (:enrollmentUids)");
+      sqlParams.addValue("enrollmentUids", UID.toValueList(params.getEnrollments()));
+    }
+
+    if (params.hasEnrollmentStatus()) {
+      sql.append(hlp.whereAnd()).append("e.status = :enrollmentStatus");
+      sqlParams.addValue("enrollmentStatus", params.getEnrollmentStatus().name());
+    }
+
+    if (params.hasFollowUp()) {
+      sql.append(hlp.whereAnd()).append("e.followup = :followUp");
+      sqlParams.addValue("followUp", params.getFollowUp());
+    }
+
+    if (!params.isIncludeDeleted()) {
+      sql.append(hlp.whereAnd()).append("e.deleted = false");
+    }
+  }
+
+  private void addTrackedEntityConditions(
+      StringBuilder sql,
+      EnrollmentQueryParams params,
+      MapSqlParameterSource sqlParams,
+      SqlHelper hlp) {
+    if (params.hasTrackedEntity()) {
+      sql.append(hlp.whereAnd()).append("te.uid = :trackedEntityUid");
```

Comment:
When we are mapping to query params we are doing ACL on the tracked entity, this is why we need to get it from the DB at the moment of the mapping. Is it necessary or are we able to have all our ACL check in the query here?

Probably nothing to fix now, I am just commenting so we can discuss it

Reply:
The TE is an optional field in the request.
The ideal scenario for us is when it's specified, because then we can run the ACL before the query.
If it's not, then we have the ACL check after we get the DB results. This is currently breaking pagination, so we need to move it into the query.

As mentioned somewhere else in this PR, it'd be great to decide if we want to make the TE param mandatory in this endpoint, as that would make the ACL check easier to implement.

As a side note, we'd still need to get the TE UID, as that's expected in the API response. What I think we won't need to return is the TET UID and TET sharing object.